spec/unit/job_fetcher_spec.rb

```diff
@@ -138,6 +138,62 @@ module Backend
             end
           end
         end
+
+        context 'when `wait` is set to true' do
+          let(:enqueued_job) { [job_queue, subject.encode(BackgroundJob.new)] }
+
+          it 'calls a blocking redis command' do
+            expect(test_redis).to receive(:blpop)
+            subject.fetch
+          end
+
+          context 'and `max` is nil' do
+            it 'returns one job' do
+              expect(test_redis).to receive(:blpop).and_return(enqueued_job)
+              result = subject.fetch
+
+              expect(result.class).to eq Resque::Job
+            end
+          end
+
+          context 'and `max` is not nil' do
+            it 'returns one job' do
+              expect(test_redis).to receive(:blpop).and_return(enqueued_job)
```

Comment:
I doubt  the utility of such tests. Later on we may want to upgrade to `BLMPOP` and this test will fail. So it only tests that the implementation is what we have implmented but doesn't validate the behavior is what we want in case of an upgrade.

That's the reason I don't write such tests.

I can imagine an integration test, where `fetch` is called, then we insert a record and validate what we want is returned, but then how do we make sure that method was called only once...

Perhaps there is value, to make whoever updates the code later to think whether the new implementation is appropriate. So I guess good to stay. Some things are not really feasible to test though :/ same goes to the other sode about `lpop`, etc.

Reply:
I think unit tests are never superfluous. They ensure the method behavior is correct, which is a safety measure when refactoring. When making changes on the code, it's normal to sometimes have to fix tests because they became invalid.

Reply:
My take is that test is good when it tests the behavior that matters. I don't care what method is called as long as the desired behavior is met.

Here we test that `blpop` is being called. But then after 2 weeks I decide to refactor this and of course I have no recollection whatsoever that I've ever seen this code, so change it to `lpop`. Then the test fails because I'm not calling `blpop`. My reaction is to update the test with `lpop`.

But then the behavior would be incorrect.

If the test doesn't tell me somehow that I need to be blocking and I don't check that it blocks manually somehow, then this test didn't stop me from introducing an issue.

If the test could catch whether I'm blocking instead of checking for the specific method name, then if I update the code using `blmpop` the test will pass. And if I use `lpop`, the test would fail. So showing me usefully that behavior is or is not the intended one.

In the current form, the test relies on my personal discretion and I wouldn't recommend that 8-)

Reply:
The behavior being tested here is `when 'wait' is set to true and 'max' is not nil it returns one job`. So when updating the test after changes, that's the behavior you should ensure this test is testing. The method used to verify it returns only one job doesn't matter as long as it actually verifies it returns one job. It also doesn't matter if this particular test blocks or not, because the test is not testing the blocking behavior.

Above there's a test that verifies the blocking behavior (`when 'wait' is set to true it calls a blocking redis command`). Your comment makes more sense applied to that test, because the blocking behavior is verified by expecting a call to `blpop`, but I couldn't find a better way to test that. If somebody updates the test to make it expect a call to `lpop` the test would pass and that would be incorrect, but I expect whoever updates the test in the future to read the test title and understand what is it trying to test before making changes to it.